Prepare the model player for the immersive presentation

jeanhbr opened the pull request:
#### 42a042b2f09f8d9f1ce03936a13be5360df66b41
<pre>
Prepare the model player for the immersive presentation
<a href="https://bugs.webkit.org/show_bug.cgi?id=303400">https://bugs.webkit.org/show_bug.cgi?id=303400</a>
<a href="https://rdar.apple.com/165702118">rdar://165702118</a>

Reviewed by Etienne Segonzac.

Add logic in the model player to prepare for an immersive presentation.
Ensure the model player is ready and the model is loaded before allowing
the immersive presentation.
Add the exitImmersive plumbing to exit an immersive presentation.

* LayoutTests/model-element/immersive/model-element-immersive-basic-expected.txt:
* LayoutTests/model-element/immersive/model-element-immersive-basic.html:
* LayoutTests/model-element/immersive/model-element-immersive-hidden-inline-expected.txt:
Test results updated.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::ensureImmersivePresentation):
(WebCore::HTMLModelElement::exitImmersivePresentation):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::ensureImmersivePresentation):
(WebCore::ModelPlayer::exitImmersivePresentation):
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/Modules/model-element/ModelPlayerTransformState.h:
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.cpp:
(WebCore::PlaceholderModelPlayer::ensureImmersivePresentation):
(WebCore::PlaceholderModelPlayer::exitImmersivePresentation):
Invalidate the transform if an exit immersive transition is performed on
a placeholder model player.

* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.h:
* Source/WebCore/dom/DocumentImmersive.cpp:
(WebCore::DocumentImmersive::exitImmersive):
(WebCore::DocumentImmersive::requestImmersive):
(WebCore::DocumentImmersive::updateElementIsImmersive):
* Source/WebCore/dom/DocumentImmersive.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::createLayer):
(WebKit::computeSRT):
Compute the tranform for an immersive presentation - scaled for the physical
world.

(WebKit::ModelProcessModelPlayerProxy::computeTransform):
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::didFailLoading):
(WebKit::ModelProcessModelPlayerProxy::sizeDidChange):
(WebKit::ModelProcessModelPlayerProxy::setEnvironmentMap):
(WebKit::ModelProcessModelPlayerProxy::applyEnvironmentMapDataAndRelease):
(WebKit::ModelProcessModelPlayerProxy::ensureImmersivePresentation):
(WebKit::ModelProcessModelPlayerProxy::exitImmersivePresentation):
(WebKit::ModelProcessModelPlayerProxy::setImmersivePresentation):
(WebKit::ModelProcessModelPlayerProxy::ensureModelLoaded):
(WebKit::ModelProcessModelPlayerProxy::triggerModelLoadedCallbacks):
Ensure the model is fully loaded before completing the immersive transition.

* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::ensureImmersivePresentation):
(WebKit::ModelProcessModelPlayer::exitImmersivePresentation):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerTransformState.cpp:
(WebKit::ModelProcessModelPlayerTransformState::setHasPortal):
(WebKit::ModelProcessModelPlayerTransformState::setStageMode):
(WebKit::ModelProcessModelPlayerTransformState::invalidateTransform):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerTransformState.h:

Canonical link: <a href="https://commits.webkit.org/303901@main">https://commits.webkit.org/303901@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04b839b08dc451dca742882c53395356a8fdd161

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133853 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6364 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45044 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141430 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85913 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135723 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6893 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6228 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102394 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/69704 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136800 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4861 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119996 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83193 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4740 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2357 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113894 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38113 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144076 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6034 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38692 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110763 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6116 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5146 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110964 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28159 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4593 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116250 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59781 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6086 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34530 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5932 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69550 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6177 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6040 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/Modules/model-element/ModelPlayer.h:
```diff
@@ -26,7 +26,9 @@
 
 #pragma once
 
+#include <WebCore/ExceptionOr.h>
```
You don't seem to be using ExceptionOr in the header. Is this needed?

jeanhbr replied:
It is not. Thanks for catching this!

etiennesegonzac commented on Source/WebCore/Modules/model-element/PlaceholderModelPlayer.cpp:
```diff
@@ -286,4 +286,19 @@ ModelPlayerAccessibilityChildren PlaceholderModelPlayer::accessibilityChildren()
+    completion(std::nullopt);
+}
+void PlaceholderModelPlayer::exitImmersivePresentation(CompletionHandler<void()>&& completion)
+{
```
shouldn't we ASSERT_NOT_REACHED here too?
(and not invalidate transform, keeping it sub-by)

etiennesegonzac commented on Source/WebCore/Modules/model-element/ModelPlayer.cpp:
```diff
@@ -209,4 +209,19 @@ void ModelPlayer::resetModelTransformAfterDrag()
+
+void ModelPlayer::exitImmersivePresentation(CompletionHandler<void()>&& completion)
+{
+    completion();
```
same question

etiennesegonzac commented on LayoutTests/model-element/immersive/model-element-immersive-basic-expected.txt:
```diff
@@ -1,13 +1,13 @@
-FAIL CSS :immersive pseudo-class matches correctly promise_test: Unhandled rejection with value: object "TypeError: Model Player created, but next steps are not implemented"
-FAIL Model is complete once presented as immersive promise_test: Unhandled rejection with value: object "TypeError: Model Player created, but next steps are not implemented"
-FAIL Multiple concurrent requestImmersive calls are handled correctly promise_test: Unhandled rejection with value: object "TypeError: Model Player created, but next steps are not implemented"
+FAIL CSS :immersive pseudo-class matches correctly promise_test: Unhandled rejection with value: object "TypeError: Not fully implemented, model not displaying in browser."
```
can we add a couple of try/catch in the test to see the pseudo-class tests turn PASS?

etiennesegonzac commented on Source/WebCore/dom/DocumentImmersive.cpp:
```diff
@@ -142,6 +149,23 @@ void DocumentImmersive::exitRemovedImmersiveElement(HTMLModelElement* element)
+{
+    if (isImmersive) {
+        if (auto oldElement = m_immersiveElement.get()) {
+            oldElement->exitImmersivePresentation( [] { } );
```
the space after `(` and before `)` look weird, is `check-webkit-style` happy with this?
(same for the other)
grepping around I think the style is usually `exitImmersivePresentation([]{ })`

etiennesegonzac commented on Source/WebCore/dom/DocumentImmersive.cpp:
```diff
@@ -142,6 +149,23 @@ void DocumentImmersive::exitRemovedImmersiveElement(HTMLModelElement* element)
     document().scheduleRenderingUpdate(RenderingUpdateStep::Immersive);
 }
 
+void DocumentImmersive::setImmersiveElement(HTMLModelElement* element, bool isImmersive)
```
could we just infer `isImmersive == false` from this being called with a nullptr element?
and invalidate the style for `m_immersiveElement` before resetting it

etiennesegonzac commented on Source/WebCore/dom/DocumentImmersive.cpp:
```diff
@@ -142,6 +149,23 @@ void DocumentImmersive::exitRemovedImmersiveElement(HTMLModelElement* element)
+void DocumentImmersive::setImmersiveElement(HTMLModelElement* element, bool isImmersive)
+{
+    if (isImmersive) {
+        if (auto oldElement = m_immersiveElement.get()) {
```
`if (RefPtr oldElement...`

etiennesegonzac commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -431,15 +431,19 @@ static void computeScaledExtentsAndCenter(simd_float2 boundsOfLayerInMeters, sim
 }
 
-static RESRT computeSRT(CALayer *layer, simd_float3 originalBoundingBoxExtents, simd_float3 originalBoundingBoxCenter, float boundingRadius, bool isPortal, CGFloat pointsPerMeter, WebCore::StageModeOperation operation, simd_quatf currentModelRotation)
+static RESRT computeSRT(CALayer *layer, simd_float3 originalBoundingBoxExtents, simd_float3 originalBoundingBoxCenter, float boundingRadius, bool isPortal, CGFloat pointsPerMeter, WebCore::StageModeOperation operation, simd_quatf currentModelRotation, bool realWorldTransform)
```
nit: `useRealWorldTransform`

etiennesegonzac commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -653,7 +661,12 @@ static RESRT modelLocalizedTransformSRT(RESRT originalSRT)
-    applyEnvironmentMapDataAndRelease();
+    applyEnvironmentMapDataAndRelease([weakThis = WeakPtr { *this }] () mutable {
+#if ENABLE(MODEL_ELEMENT_IMMERSIVE)
+    if (weakThis)
```
you know the drill :) (`RefPtr protectedThis`...)

etiennesegonzac commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -1017,6 +1046,67 @@ static void setIBLAssetOwnership(const String& attributionTaskID, REAssetRef ibl
+void ModelProcessModelPlayerProxy::setImmersivePresentation(bool immersivePresentation)
+{
+    [CATransaction begin];
+    [CATransaction setDisableActions:YES];
```
I think we might want to set the `WebActionDisablingCALayerDelegate` on this layer instead

etiennesegonzac commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
```diff
@@ -203,6 +208,16 @@ class ModelProcessModelPlayerProxy final
+
+#if ENABLE(MODEL_ELEMENT_IMMERSIVE)
+    bool m_immersivePresentation { false };
+    WebCore::LayoutSize m_inlineFrameSize { };
```
why not m_layoutSize? :) (we only "layout" on the page so the inline is kinda implied)

jeanhbr committed:
Prepare the model player for the immersive presentation
https://bugs.webkit.org/show_bug.cgi?id=303400
rdar://165702118

Reviewed by Etienne Segonzac.

Add logic in the model player to prepare for an immersive presentation.
Ensure the model player is ready and the model is loaded before allowing
the immersive presentation.
Add the exitImmersive plumbing to exit an immersive presentation.

* LayoutTests/model-element/immersive/model-element-immersive-basic-expected.txt:
* LayoutTests/model-element/immersive/model-element-immersive-basic.html:
* LayoutTests/model-element/immersive/model-element-immersive-hidden-inline-expected.txt:
Test results updated.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::ensureImmersivePresentation):
(WebCore::HTMLModelElement::exitImmersivePresentation):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::ensureImmersivePresentation):
(WebCore::ModelPlayer::exitImmersivePresentation):
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/Modules/model-element/ModelPlayerTransformState.h:
* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.cpp:
(WebCore::PlaceholderModelPlayer::ensureImmersivePresentation):
(WebCore::PlaceholderModelPlayer::exitImmersivePresentation):
Invalidate the transform if an exit immersive transition is performed on
a placeholder model player.

* Source/WebCore/Modules/model-element/PlaceholderModelPlayer.h:
* Source/WebCore/dom/DocumentImmersive.cpp:
(WebCore::DocumentImmersive::exitImmersive):
(WebCore::DocumentImmersive::requestImmersive):
(WebCore::DocumentImmersive::updateElementIsImmersive):
* Source/WebCore/dom/DocumentImmersive.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::createLayer):
(WebKit::computeSRT):
Compute the tranform for an immersive presentation - scaled for the physical
world.

(WebKit::ModelProcessModelPlayerProxy::computeTransform):
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::didFailLoading):
(WebKit::ModelProcessModelPlayerProxy::sizeDidChange):
(WebKit::ModelProcessModelPlayerProxy::setEnvironmentMap):
(WebKit::ModelProcessModelPlayerProxy::applyEnvironmentMapDataAndRelease):
(WebKit::ModelProcessModelPlayerProxy::ensureImmersivePresentation):
(WebKit::ModelProcessModelPlayerProxy::exitImmersivePresentation):
(WebKit::ModelProcessModelPlayerProxy::setImmersivePresentation):
(WebKit::ModelProcessModelPlayerProxy::ensureModelLoaded):
(WebKit::ModelProcessModelPlayerProxy::triggerModelLoadedCallbacks):
Ensure the model is fully loaded before completing the immersive transition.

* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::ensureImmersivePresentation):
(WebKit::ModelProcessModelPlayer::exitImmersivePresentation):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerTransformState.cpp:
(WebKit::ModelProcessModelPlayerTransformState::setHasPortal):
(WebKit::ModelProcessModelPlayerTransformState::setStageMode):
(WebKit::ModelProcessModelPlayerTransformState::invalidateTransform):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerTransformState.h:

Canonical link: https://commits.webkit.org/303901@main
